lib/Web/Server.hs

```diff
@@ -0,0 +1,249 @@
+-- SPDX-FileCopyrightText: 2022 Serokell <https://serokell.io>
+--
+-- SPDX-License-Identifier: MPL-2.0
+
+module Web.Server where
+
+import Control.Monad.Catch (SomeException, try)
+import Control.Monad.Except
+
+import Data.ByteString.Lazy qualified as Bytes.Lazy
+import Data.Function ((&))
+import Data.HashMap.Strict (HashMap)
+import Data.HashMap.Strict qualified as HashMap
+import Data.Text (Text)
+import Data.Text qualified as Text
+import Data.Text.Encoding qualified as Text
+
+import Polysemy
+import Polysemy.Error hiding (try)
+
+
+import Backend
+import Backend.Vault.Kv
+import CLI.Types
+import Entry
+import Error
+import Web.API
+
+import Backend.Interpreter (runBackend)
+import Coffer.Directory (Directory)
+import Coffer.Path (EntryPath, Path, QualifiedPath)
+import Data.Functor ((<&>))
+import Data.Set qualified as Set
+import Fmt (pretty)
+import Servant.API
+import Servant.Client
+import Servant.Server
+
+url :: BaseUrl
+url = BaseUrl Http "127.0.0.1" 8200 ""
+
+mount :: Text
+mount = "secret"
+
+runBackendIO' :: Sem '[BackendEffect, Error CofferError, Embed IO, Final IO] a -> IO (Either CofferError a)
+runBackendIO' action =
+  runBackend action
+    & errorToIOFinal @CofferError
+    & embedToFinal @IO
+    & runFinal
+
+reportErrors :: IO (Either CofferError a) -> Handler a
+reportErrors io = do
+  eea <- liftIO do
+    try @_ @SomeException io
+
+  case eea of
+    Left ioError -> do
+      throwError err500 { errBody = errorToServantErrorMsg ioError }
+
+    Right (Left e) -> do
+      throwError err500 { errBody = errorToServantErrorMsg (pretty @_ @Text e) }
+
+    Right (Right a) -> do
+      return a
+
+errorToServantErrorMsg :: Show a => a -> Bytes.Lazy.ByteString
+errorToServantErrorMsg
+  = Bytes.Lazy.fromStrict
+  . Text.encodeUtf8
+  . Text.pack
+  . show
+
+makeServer
+  :: (forall a. VaultToken -> Command a -> Handler a)
+  -> Server API
+makeServer run token
+  =    view   run token
+  :<|> create run token
+  :<|>
+    (\txt fkey ->
+         private run token txt fkey
+    :<|> public  run token txt fkey
+    :<|> set     run token txt fkey
+    )
+  :<|> deleteField run token
+  :<|> find'       run token
+  :<|> rename      run token
+  :<|> copy'       run token
+  :<|> delete'     run token
+  :<|> tag         run token
+
+view
```

Comment:
I'm not sure, but could the common repeating code
```
_ run token _ = run token _
``` 
be extracted and written once? (In the next group of functions)

Reply:
I think it wouldn't type check

Reply:
Ah, I see